### PR TITLE
WC2-171 Entities Deduplication Feature

### PR DESCRIPTION
**Pull request by mathvdh**
Give a way to the users to find duplicated entities (per same entity type and account)
Allow them to run algorithms/analyzes that will find the duplicates (by choosing which fields on the reference form of the entity type to analyze + other params for the algorithm)
List the found duplicates and allows the users to ignore or merge them interactively.

Related JIRA tickets : WC2-171, WC2-137, WC2-230, WC2-231, WC2-236

## Self proofreading checklist

- [x] Did I use eslint and black formatters
- [x] Is my code clear enough and well documented
- [x] Are my typescript files well typed
- [x] New translations have been added or updated if new strings have been introduced in the frontend
- [x] My migrations file are included
- [x] Are there enough tests
- [ ] Documentation has been included (for new feature)

## Changes

Backend + Background tasks and frontend for the feature.
One implementation of an algorithm to find duplicates.

## How to test

In order to test locally you will need to have at least one Entity Type setup correctly as well as multiple entities with the same or similar attributes (reference form of the Entity Type) filled. You can find in `iaso/tests/fixtures/test_form_deduplication.xlsx` a simple form with 3 fields : Prenom, Nom, Age that you can use (see the test code in `iaso/tests/api/deduplication/test_entities_deduplication.py`

After that you need to create an initial analyze configuration using the API, you need to send a POST request to the URL `/api/entityduplicates_analyzes/` with a JSON payload similar to this one (depending on the fields you have in your reference form) :

```json
{
  "fields": [
    "What_is_the_father_s_name",
    "What_is_the_child_s_name"
  ],
  "parameters": {},
  "entity_type_id": "2"
}
```

If you are in local, You need to have a background queue job worker running and it can be launched using `./manage.py tasks_worker` or `docker-compose exec iaso ./manage.py tasks_worker` if you are in docker.

After that if there are similar entities, entity duplicates will be created and you can use the frontend to see and ignore/merge them use the dashboard menu Beneficiaries > Duplicates or visit this url : `/dashboard/entities/duplicates/accountId/2/order/id/pageSize/20/page/1`

## Print screen / video

Screenshot 2023-06-09 at 12 19 02
Screenshot 2023-06-09 at 12 16 00
Screenshot 2023-06-09 at 12 09 28

### CODE REVIEW

File: iaso/models/deduplication.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
+from django.db import models
+from iaso.models import Entity, EntityType, Task
+import iaso.models.base as base_models
+from iaso.api.deduplication.algos import POSSIBLE_ALGORITHMS, DEFAULT_ALGORITHM  # type: ignore
+from django.utils.timezone import now
+
+PENDING = "PENDING"
+VALIDATED = "VALIDATED"
+IGNORED = "IGNORED"
+VALIDATION_STATUS = [(PENDING, "Pending"), (VALIDATED, "Validated"), (IGNORED, "Ignored")]
```
Comment (olethanh) [inline iaso/models/deduplication.py]: please use django builtin TextChoice. cf https://docs.djangoproject.com/en/3.2/ref/models/fields/#enumeration-types

File: iaso/models/deduplication.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+DUPLICATE = "DUPLICATE"
+COUSIN = "COUSIN"
+PRODUCED = "PRODUCED"
+DUPLICATE_TYPES = [(DUPLICATE, "Duplicate"), (COUSIN, "Cousin"), (PRODUCED, "Produced")]
```
Comment (olethanh) [inline iaso/models/deduplication.py]: idem use please use django builtin TextChoice. cf https://docs.djangoproject.com/en/3.2/ref/models/fields/#enumeration-types

File: iaso/api/deduplication/algos/finalize.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+from django.utils.timezone import now
+from django.apps import apps
+
+
+def finalize_from_task(the_task: Task, potential_duplicates: List[PotentialDuplicate]):
+    eda = the_task.entity_duplicate_analyze.first()
+    if not eda:
+        raise Exception("No entity duplicate analyze found for task %s" % the_task)
+
+    ed_model = apps.get_model("iaso", "EntityDuplicate")
```
Comment (olethanh) [inline iaso/api/deduplication/algos/finalize.py:14]: why not just import it?

File: iaso/models/deduplication.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+
+class EntityDuplicateAnalyze(models.Model):
```
Comment (olethanh) [inline iaso/models/deduplication.py]: Analysis

**Review by olethanh — APPROVED** [review]

File: iaso/models/deduplication.py (hunk 1 above)
Comment (mathvdh) [reply iaso/models/deduplication.py]: Done

File: iaso/models/deduplication.py (hunk 2 above)
Comment (mathvdh) [reply iaso/models/deduplication.py]: Done

File: iaso/api/deduplication/algos/finalize.py (hunk 3 above)
Comment (mathvdh) [reply iaso/api/deduplication/algos/finalize.py:14]: Circular import ...

**Comment by hakifran** [comment]
@mathvdh this PR has a conflict in iaso/admin.py

**Review by hakifran — APPROVED** [review]

**Review by codeKameleon — APPROVED** [review]

**Pull request merged**